Add maintainer's guide

Riksu9000 opened the pull request:
This document clarifies how the maintainers should manage the project.

We should write about using the milestones as well, but I want to discuss that first. I'll open an issue about that. If that doesn't get resolved in time, it can be another PR. EDIT: Here's the issue #1515

I've added new instructions on handling stale PRs. Three months is quite conservative, as it concerns any activity, and can be reduced. Let me know what you think.

JF002 commented on doc/maintainer-guide.md:
```diff
@@ -0,0 +1,30 @@
+Approving a PR means that the reviewer has deemed the PR ready to be merged.
+
+While reviewing
+
```
I would add 2 points here:
 - Questions whether the new functionality is needed in the project and if it follows the vision of the projects ("Do we really want/need that feature?").
 - Questions whether the maintainer will be willing to maintain this feature on the long term ("Do we have the resources to maintain this feature?")

JF002 commented on doc/maintainer-guide.md:
```diff
@@ -0,0 +1,30 @@
+and when the author is specified as the maintainer of the code in question.
+Then only a single approval is needed.
+
+Commits that fix code formatting from previous commits must be squashed before merging a PR,
```
I think we could ask the author of the PR to make sure the history is clean before we merge the PR, indeed!

Riksu9000 replied:
We haven't had any requirements for commits so far, and this is an idea I had to have just slightly cleaner commits. Requirements for commits is probably something worth discussing, as it relates to #1515, so I think I'll open a new issue for that as well.

Riksu9000 replied:
I think rebase merging gets rid of the merge commits anyway, so a linear history is already implicitly enforced by this guide. And we should definitely enforce a linear history. https://docs.github.com/en/github/administering-a-repository/about-protected-branches#require-linear-history

JF002 reviewed (commented):
This is a great initiative!

Riksu9000 commented:
I modified the merging section. I removed the part about the author being specified as the maintainer of the code, as that's not something we do currently, and replaced it with allowing small maintenance PRs by core developers to be merged with a single approval. This should reduce the wait times for maintenance PRs.

JF002 commented:
Looks good to me. That's a good start, and we'll probably refine it later on :)

Riksu9000 commented on doc/maintainer-guide.md:
```diff
@@ -0,0 +1,37 @@
+or the PR must be squash merged.
+Otherwise the PR shall be squash or rebase merged.
+
+All but the simulator check must be passed before merging a PR.
```
@JF002 There are branch protection rules which make it so checks must be passed before merging. If these are enabled for build-firmware, clang-format and clang-tidy, this line can be removed.

JF002 replied:
Do you mean this for the **develop** branch? 
![image](https://user-images.githubusercontent.com/2261652/212159133-a8cf0fae-eb35-4369-b653-29256202ea9e.png)

Enabling as many auto checks as possible is probably a good idea to make our lifes easier, but I also think it's a good idea to mention those checks in the maintainer's guide, as a documentation of the process.

Riksu9000 replied:
Okay, I'll keep the note here then. I'll also keep the wording, even though it doesn't exactly match the branch protection rules.

JF002 commented:
Regarding the merging section, and since @Riksu9000 brought to my attention that we can enable protection rules on branches, do we want to add a "linear history" rule to this guide, and enforce it with this check ?

![image](https://user-images.githubusercontent.com/2261652/212160322-6de1b83c-8813-4fb0-8cea-f21cde38a759.png)

Riksu9000 commented:
I think we should enable the branch protection rule, but I'm not sure we need to explicitly mention it in the guide. Following the instructions in the guide will maintain a linear history, and the branch protection rule makes it so you couldn't break the rule anyway. I'll be updating the merging section soon.
